### DISCUSSION TITLE
2024-07-18 - Awesome READMEs - Chewing the Fat

### DISCUSSION THREAD

**Discussion by sethdaily** [SSWConsulting/SSW.Rules.Content#8907] [Chewing the Fat]
Hey documentation experts,

Let's talk about https://www.ssw.com.au/rules/awesome-readme/

1. What's the best README you've ever seen?

2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

**Comment by JeanThirion** [comment]
I've seen some real bad README's before... But nowadays they're pretty good at SSW.

Can't we get someone smart to write the README (or part of it) based on the Repo (TODOs etc...), Discussions and issues ? I'm sure we could put something together. Not sure it'd be useful but that'd be fun.

--- 

> What's the best README you've ever seen?

Impossible to tell but I'm generally speaking quite proud of SSWs

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - I think 3 templates is a great idea.

**Comment by Freego1783** [comment]
> What's the best README you've ever seen?

Hard to tell, I have probably seen good ones, but not one good enough to remember it easily 

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - I'g do with 3 templates. I think it is nice to have multiple options depending on the project.

**Comment by tiagov8** [comment]
> What's the best README you've ever seen?

Not sure, but I like SSW.Rules.Content combined with the Wiki

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - I think it's a good starting point to enforce projects to have them.

--

My 2 cents:

I tend to think the INSTRUCTIONS / HOW TOs / KNOWN ISSUES are more suitable into the **Wiki** instead... not a strong opinion though

**Comment by matt-goldman-ssw** [comment]
>What's the best README you've ever seen?

Hard to say, but these two are probably close to the top:
https://github.com/langflow-ai/langflow
https://github.com/git-ecosystem/git-credential-manager

>Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Strongly agree

**Comment by Levijj22** [comment]
Should we build a GPT to write README's?
This would help with inconsistency.
For existing files, we could train it to evaluate and 'beef' up.

Not sure about the idea of having 3 types of README types.
Maybe we could have a 5-star rating system built in that devs would vote on when they used it.
Then you know what to expect if it's 1-star etc.

**Comment by ncn-ssw** [comment]
> 1. What's the best README you've ever seen?

I cannot recall 1 that I think of as the "best" - but it is certainly easy to see when they are bad.

> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think only 1 template should be created with the following items.

1. H1/H2 titles
2. Intro paragraph
3. Diagram
4. Installation instructions for users
5. Installation instructions for contributors
6. Contributor expectations (if public)

I think including "Known Issues" risks the document quickly falling out of date.

**Comment by tkapa** [comment]
I only ever recall interacting with bad README's because the friction to get started is high. Good README's give you the info you want at a glance, and you can easily forget about it.

We should just have one template that things can be removed from.

**Reply by bradystroud** [reply]
> I only ever recall interacting with bad README's because the friction to get started is high. Good README's give you the info you want at a glance, and you can easily forget about it.

Great point

**Comment by leahy268** [comment]
>What's the best README you've ever seen?
Not sure of any specific ones.  But usually a diagram of how the program works.  Good info on how to use it. 

>Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Maybe.  But like any template they can easily be ignored.

**Comment by Tanya-Leahy** [comment]
>What's the best README you've ever seen?
I haven't used a README yet but I am finding information in Wiki's really useful.

>Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Maybe.

**Comment by stephanfakossw** [comment]
N/a - not a dev, but I can see how this would be important to devs

**Comment by calumjs** [comment]
We should be able to create a custom GPT to generate readmes fairly easily.
Not sure if I agree with known issues going in there, unless you have no intention of solving them.
And not sure if SSW repos should have donation links.

**Comment by duanxinhuan** [comment]
Hi @adamcogan, can we send a email to MS and tell them they did a terrible job on documentation, they should follow golden readme rules :)

**Comment by jimmy1992abc** [comment]
I am not a dev, I feel like the README is similar to the patch note that I saw in other software.
The patch note always follows the same strcuture in the same order each time, so I know which part I expect and which section I can skip.
For README it could be the same.

**Comment by ricksu978** [comment]
> What's the best README you've ever seen?

N/A, not good enough for me to remember it.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes.
I remember a client has repo creation process integrated to their Slack channel with the bot.
Type a command in the chat and then answer few questions, a brand new repo will be created for you.

**Comment by joshbermanssw** [comment]
>What's the best README you've ever seen?

N/A - not memorable enough

>Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - I like the distinction, although i think the gold and silver should be closer than silver and bronze

**Comment by Calinator444** [comment]
> 1. What's the best README you've ever seen?

In my (somewhat biased) opinion, the SSW Website has a very good readme. There's a lot of moving parts involved, and I think the readme does a good job of grouping everything into headings.

> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

The Gold Silver template sound like a good idea. Not sure how I fell about the bronze template. A good readme is like Good UX design, it's important to have informative up to date instructions for setting up the project regardless of whether it's for public facing or for internal projects.

**Comment by edkranz** [comment]
> What's the best README you've ever seen?

READMEs serve purely a utilitarian purpose, so you only remember the bad ones that fail to reach this benchmark. If there's anything memorable in a README then it probably doesn't need to be in a README.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes! Standardising documentation is always a good idea, and the README is the purest/most prolific form of documentation for any software project. Having different tiers of READMEs for serving different purposes is a good idea, of course you may need to make a user install video for consumer software, but you obviously wouldn't for a small internal utility library, for example.

**Comment by bettybondoc** [comment]
> What's the best README you've ever seen?

Can't remember

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - Using these tiers will make the task less daunting and, in turn, more likely to get done. Categorising the READMEs lets us focus our efforts, avoid overwhelming documentation for smaller projects, and keep everything consistent and high-quality.

**Comment by awaltos** [comment]
> What's the best README you've ever seen?

The only one that comes to mind is for a [UI library](https://github.com/aws-amplify/amplify-ui/blob/main/README.md) that I used. Not sure if it's the best - but I liked how it didn't cram everything into 1 file and used links instead.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Template is a good idea, but maybe just the one. It can have all the important sections and teams can decide which sections need to be included/removed per project. If a "gold" project needs all-singing and all-dancing then it probably needs a wiki or website or multiple readmees

**Comment by GordonBeeming** [comment]
I love a good readme... but it shouldn't contain all the verbose information suggested in this topic. Links off to some of the docs is better. 
For internal project, having ADRs, DoR and DoD links, as well as environment information so everyone knows how to access dev/test/prod urls is probably more valuable front and center

---------

> What's the best README you've ever seen?

Not sure on the best ever... you sort of have good experiences and harder experiences... I like the vscode readme, but not sure if it's "the best I've seen" 
https://github.com/microsoft/vscode

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

No - anytime something like this gets made into a template, the overall quality of the artifact lowers as boxes are ticked to "meet the standard"... instead of focusing on what would be valuable for the project specifically... e.g.: writing a nuget package needs different content to writing a cli tool

**Comment by Ben0189** [comment]
>1. What's the best README you've ever seen?
Vs code github repo readme

Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Yes - good to have a readme template

**Comment by bahjatalaadelssw** [comment]
> What's the best README you've ever seen?
It would a missing readme, because then you have no baseline

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Yes, but make it one doc that you can delete from to cater for the project needs

**Comment by kikibianc** [comment]
> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes, but make it flexible enough to add sections based on what's needed.

**Comment by isaaclombardssw** [comment]
>  What's the best README you've ever seen?

Not sure about this.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think it is a good idea, but there needs to be some flexibility as well – depends on the scenario as to what should really be included.

**Comment by chrisschultzssw** [comment]
> What's the best README you've ever seen?

Can't think of a specific one.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes, I think there should be one template - with optional sections for internal/PoC projects.

**Comment by jaydenalchin** [comment]
As long as a readme is clear, structured, and understandable it's a win. 

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think a template is a great idea - but we don't need 3 of them when we can capture core requirements + optional nice-to-haves all in one. Less is more 😉

**Comment by Marxoz** [comment]
> What's the best README you've ever seen?
The one at my last client was pretty good. I know, cause I had to set up my project a couple of times. Also, it was mostly completed by SSW folks

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Yes, but just one. You can always delete sections if it is not relevant.

**Comment by AttackOnMorty** [comment]
> What's the best README you've ever seen?

N/A

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I prefer one template. We should establish a baseline but it should be flexible.

**Comment by jeoffreyfischer** [comment]
> What's the best README you've ever seen?

N/A

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I'd rather have ✅ **3 light templates** rather than ❌ **1 big template** containing lots of "if" e.g. "If this is for a client..." "If this is for internal project...", "If this is for a quick and dirty project..." etc.

**Comment by christoment** [comment]
> 1. What's the best README you've ever seen?

Can't think of one specifically, but well-known project like dotnet/aspnetcore and Angular have good READMEs.

> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes, but not tiered like Gold, Silver and Bronze - different purpose should have different template (e.g. I wouldn't put cloud architecture diagram on my public facing readme)

**Comment by jerryluossw** [comment]
> What's the best README you've ever seen?

[https://github.com/microsoft/chat-copilot](https://github.com/microsoft/chat-copilot)

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes, it helps a lot, and it's better we only create 1 template with some required and optional sections.

**Comment by Dhruv-0987** [comment]
What's the best README you've ever seen?
Don't think there is right answer to this depend on project to project.
Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Yes very good idea, every dev just wants to fill in the blanks

**Comment by MattParkerDev** [comment]
> What's the best README you've ever seen?

RulesGPT is pretty good

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

If we really feel it is a worthwhile venture.

**Comment by JackDevAU** [comment]
> What's the best README you've ever seen?

Can't decide 😂 but there are a lot of bad ones.... and they are mainly C# projects.... 

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Could be worth it! Keeps SSW projects consistent which would help everyone get started quicker.

I think not only should we be making the Readmes good but we should also be adding, 

- Code of Conduct
- License
- Security

![image](https://github.com/user-attachments/assets/1db35ac2-a50f-4d11-a36f-e20d6b0b5356)
**Figure: Other Important files**

### 🥇 Gold README
```
Title

Build Info Embeds 

Bullet points / Features

Demo

Docs link 

Contributing / Installation 

Usage (Depends what this project is)

Backlink to SSW or other projects
```

### 🥈 Silver README
```
Title

Build Info Embeds 

bullet points / Features

Docs link 

Contributing / Installation 

Usage (Depends what this project is)

Backlink to SSW or other projects
```

### 🥉 Bronze README
```
Title

bullet points / Features

Demo

Contributing / Installation 

Usage (Depends what this project is)

Backlink to SSW or other projects
```

**Comment by jernejk** [comment]
I like having a good README containing all the information you need in one go.

## TL;DR;

Use AI to generate documentation like customized README (remove segments not needed and update placeholders with known data) and use generate segments of the documentation that changes over time like diagrams. We could use [AutoGen](https://microsoft.github.io/autogen/) to test F5 experience with help of GitHub Codespaces.

## Generating README

For starters, if we have 1 master README template, we could use AI to tweak it and remove irrelevant segments based on the project's needs. It would still have placeholders for the user to replace as AI won't be able to do all of the work.

For instance, we could have interaction:

```
AI: What's the name of the project or link to current repo
User: github.com/....
AI: I see the project is C# and have sample code. Would you like to have a slim README for demo repos?
User: Yes
AI: What is the purpose of this demo?
User: Showing EF Core feature for updating multiple rows without fetching data
AI: Proposed README:
....

The application seems to be standard .NET Web App and I created an F5 experience for it.
Would you like to add or modify the README?
```

## Maintaining documentation

Projects like TimePro split a lot of the documentation into multiple files, which is not necessarily bad but makes it harder to get a feel for the project quickly. I also find it hard to find a good balance between too little and too much documentation. In TimePro, it is hard to keep up with all of the required documentation changes.

It would be great, if we could verify the documentation and spot parts that might be out of date based on the code we changed.

For instance, diagrams, if there are too many diagrams, the likelihood of them going obsolete without updating becomes pretty high and can confuse people. But having no diagrams can make understanding of the project hard.

## Generating diagrams

A happy medium would be automatically generated diagrams as bigger changes happen, likely with the help of AI. All of the changes, of course, need to be reviewed by a human, and we probably want this as a script rather than a CI pipeline, but this way, we could have up-to-date diagrams for our projects. I attempted to do this for a few projects with mixed results, but some were amazing! The key to an incredible diagram is to imagine what kind of diagram would work well and what the diagram should focus on.

## Testing F5 experience

Another thing about the F5 experience is that with tools like [AutoGen](https://microsoft.github.io/autogen/), it is plausible to make an AI agent follow the instructions. In fact, it can auto-heal on each step and potentially tell us what failed and how they fixed it.

Good templates, GitHub flows, and AI can do a lot. With the rise of offline AI (Chrome is planning to have Gemini built into the browser so that we can securely use SLMs on our local machines!), we are closer than ever to being able to have AI verify not only the text we write but also try to execute instructions and check if diagrams are still up-to-date.

The template also talks about a video instead of a diagram. While the video is nice, updating it for every change is expensive unless it's AI-generated.

**Comment by KristenHu** [comment]
> What's the best README you've ever seen?

Hard to say, no README is good enough to impress me.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes, I like the 3-Tier idea

**Comment by babakamyljanovssw** [comment]
> What's the best README you've ever seen?

N/A

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think it is good idea to have starting point template for Readme

**Comment by tino-liu** [comment]
> What's the best README you've ever seen?

N/A

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think having 3 different templates is a great idea

**Comment by AnkyChe** [comment]
N/A, I'm not a dev. But by doing the chewing the fat, I find README is important for a dev to understand a new project.

**Comment by Aibono1225** [comment]
> What's the best README you've ever seen?

N/A

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I prefer 3 tiers template

**Comment by jimmidier** [comment]
> What's the best README you've ever seen?

Various but usually with product status badges, link to a demo, a short startup snippet, links to user docs and contributor docs respectively, and license.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

3 would be nice

**Comment by TobyChurches** [comment]
> What's the best README you've ever seen?

Hard to say, there are a bunch of good ones out there (also alot of bad ones 😂)

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes! If we have a full suite of Templates it would be nice. Also there wouldn't be a tonne to maintain with this kind of template

**Comment by bradystroud** [comment]
> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes we should have a template. At a minimum we should have a list of requirements and nice-to-haves for a readme.
I dont think types of metal is a good way to tier READMEs. 

It would be better have different ones based on type of project (Same as the pptx template) 
- Product e.g. SugarLearning
- Template e.g. Clean Architecture

Regardless, I expect all READMEs to become inconsistent over time.

templates should live in this repo https://github.com/SSWConsulting/SSW.GitHub.Template

**Comment by danielmackay** [comment]
> 1. What's the best README you've ever seen?

https://github.com/amantinband/error-or

> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

For sure.  I realise they should be tailored, but I think a starting point for each would be great.  I thought this was missing from this CTF.  There is so much info, in what goes in a good one, but we haven't provided any examples.

**Comment by vladislav-kir** [comment]
> 1. What's the best README you've ever seen?

I cannot think of a single best or worst once. It is generally easy to see when a README is bad. SSW README are usually good :)

> 2. Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think one template would be enough.

**Comment by PennyWalker** [comment]
I don't really see a lot of them in my position, but from what I've seen at SSW, our README's are actually quite good - the one used for the Newsletter on the Website was helpful

**Comment by steven0x51** [comment]
What's the best README you've ever seen?
lots good ones, one example: https://github.com/facebook/react

Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?
Just create Gold, and remove items when N/A

**Comment by PothieuG** [comment]
> What's the best README you've ever seen?

Hard to say... I remember that a lots of repositories related to image generation projects have great REAME.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

I think 3 templates is a great idea.

**Comment by drwharris** [comment]
If I am using a library, All that really matters to me is how is it intended to be used, with examples.

If I am deving, then I need to know how to hit F5 and debug and how to use the application.

**Comment by ozair-ashfaque** [comment]
> What's the best README you've ever seen?

It is hard to pinpoint the absolute best, but READMEs that include architectural diagrams are more effective.

> Should SSW create a README template (or maybe 3 of them being Gold, Silver and Bronze)?

Yes - it is a good idea
